File: src/modules/catalog/NexTraceOne.Catalog.Domain/Contracts/Policies/InterfaceContractPolicy.cs
using NexTraceOne.Catalog.Domain.Contracts.Enums;
using NexTraceOne.Catalog.Domain.Graph.Enums;

namespace NexTraceOne.Catalog.Domain.Contracts.Policies;

/// <summary>
/// Política estática que define quais tipos de contrato são permitidos por tipo de interface de serviço.
/// Encapsula as regras de domínio do mapeamento InterfaceType → ContractType[].
/// Complementa a ServiceContractPolicy com granularidade de interface.
/// </summary>
public static class InterfaceContractPolicy
{
    /// <summary>
    /// Mapeamento oficial: InterfaceType → ContractTypes permitidos.
    /// Interfaces com lista vazia não expõem contratos de interface pública.
    /// </summary>
    private static readonly IReadOnlyDictionary<InterfaceType, ContractType[]> _allowedContractTypes =
        new Dictionary<InterfaceType, ContractType[]>
        {
            [InterfaceType.RestApi]           = [ContractType.RestApi],
            [InterfaceType.SoapService]       = [ContractType.Soap],
            [InterfaceType.KafkaProducer]     = [ContractType.Event, ContractType.SharedSchema],
            [InterfaceType.KafkaConsumer]     = [ContractType.Event, ContractType.SharedSchema],
            [InterfaceType.GrpcService]       = [ContractType.Grpc],
            [InterfaceType.GraphqlApi]        = [ContractType.RestApi],
            [InterfaceType.WebhookProducer]   = [ContractType.RestApi, ContractType.Event],
            [InterfaceType.ZosConnectApi]     = [ContractType.RestApi],
            [InterfaceType.MqQueue]           = [ContractType.MqMessage],
            [InterfaceType.IntegrationBridge] = [ContractType.RestApi, ContractType.Soap, ContractType.Event],
            // Sem contrato de interface pública
            [InterfaceType.BackgroundWorker]  = [],
            [InterfaceType.ScheduledJob]      = [],
            [InterfaceType.WebhookConsumer]   = [],
        };

    /// <summary>
    /// Tipos de interface que exigem contrato obrigatório.
    /// </summary>
    private static readonly HashSet<InterfaceType> _mandatoryContractTypes =
    [
        InterfaceType.RestApi,
        InterfaceType.SoapService,
        InterfaceType.KafkaProducer,
        InterfaceType.GrpcService,
        InterfaceType.GraphqlApi,
        InterfaceType.ZosConnectApi,
        InterfaceType.MqQueue,
    ];

    /// <summary>
    /// Indica se um tipo de interface suporta contratos de interface pública.
    /// </summary>
    /// <param name="interfaceType">Tipo da interface.</param>
    /// <returns>
    /// <c>true</c> se o tipo permite ao menos um contrato; <c>false</c> caso contrário.
    /// </returns>
    public static bool SupportsContracts(InterfaceType interfaceType)
        => _allowedContractTypes.TryGetValue(interfaceType, out var types) && types.Length > 0;

    /// <summary>
    /// Retorna os tipos de contrato permitidos para o tipo de interface.
    /// Retorna lista vazia se a interface não suporta contratos.
    /// </summary>
    /// <param name="interfaceType">Tipo da interface.</param>
    public static IReadOnlyList<ContractType> AllowedContractTypes(InterfaceType interfaceType)
        => _allowedContractTypes.TryGetValue(interfaceType, out var types)
            ? types
            : Array.Empty<ContractType>();

    /// <summary>
    /// Verifica se um tipo de contrato é permitido para o tipo de interface.
    /// </summary>
    /// <param name="interfaceType">Tipo da interface.</param>
    /// <param name="contractType">Tipo do contrato a validar.</param>
    public static bool IsContractTypeAllowed(InterfaceType interfaceType, ContractType contractType)
        => AllowedContractTypes(interfaceType).Contains(contractType);

    /// <summary>
    /// Indica se o tipo de interface exige a existência de pelo menos um contrato vinculado.
    /// </summary>
    /// <param name="interfaceType">Tipo da interface.</param>
    /// <returns>
    /// <c>true</c> se a interface exige contrato; <c>false</c> se o contrato é opcional ou não aplicável.
    /// </returns>
    public static bool RequiresContract(InterfaceType interfaceType)
        => _mandatoryContractTypes.Contains(interfaceType);
}

File: src/frontend/src/features/contracts/shared/interfaceContractPolicy.ts
/**
 * Política de contratos por tipo de interface — espelho do InterfaceContractPolicy do backend.
 * Define quais ContractTypes são permitidos para cada InterfaceType.
 * Usado para validação no frontend antes de criar um contrato binding.
 */
import type { InterfaceType } from '../../../types';
import type { ContractTypeValue } from './constants';

/**
 * Mapeamento oficial: InterfaceType → ContractType(s) permitidos.
 * Interfaces com lista vazia não expõem contratos de interface pública.
 */
export const INTERFACE_CONTRACT_POLICY: Record<InterfaceType, ContractTypeValue[]> = {
  RestApi:          ['RestApi'],
  SoapService:      ['Soap'],
  KafkaProducer:    ['Event', 'SharedSchema'],
  KafkaConsumer:    ['Event', 'SharedSchema'],
  GrpcService:      [],  // ContractType.Grpc não exposto no frontend ainda
  GraphqlApi:       ['RestApi'],
  BackgroundWorker: [],
  ScheduledJob:     [],
  WebhookProducer:  ['RestApi', 'Event'],
  WebhookConsumer:  [],
  ZosConnectApi:    ['RestApi'],
  MqQueue:          ['MqMessage'],
  IntegrationBridge: ['RestApi', 'Soap', 'Event'],
};

/**
 * Indica se um tipo de interface suporta contratos.
 */
export function supportsContractsForInterface(interfaceType: InterfaceType): boolean {
  const allowed = INTERFACE_CONTRACT_POLICY[interfaceType];
  return Array.isArray(allowed) && allowed.length > 0;
}

/**
 * Retorna a lista de tipos de contrato permitidos para um tipo de interface.
 */
export function allowedContractTypesForInterface(interfaceType: InterfaceType): ContractTypeValue[] {
  return INTERFACE_CONTRACT_POLICY[interfaceType] ?? [];
}

/**
 * Indica se um tipo de interface requer contrato por política.
 * Por padrão, qualquer interface que suporte contratos é elegível mas não obrigatória
 * a nível de política — a obrigatoriedade é definida pelo campo requiresContract na entidade.
 */
export function requiresContractForInterface(interfaceType: InterfaceType): boolean {
  return supportsContractsForInterface(interfaceType);
}
